File: backend/app/schemas/sql_compiler.py
from typing import Any, Literal

from pydantic import BaseModel, Field


SupportedDialect = Literal[
    "mysql",
    "postgresql",
    "oracle",
    "mssql",
]


class SQLCompilerRequest(BaseModel):
    prompt: str = Field(
        min_length=2,
        max_length=2000,
    )

    domain_id: int | None = None

    requested_limit: int = Field(
        default=100,
        ge=1,
        le=1000,
    )

    maximum_entities: int = Field(
        default=6,
        ge=1,
        le=15,
    )

    maximum_path_depth: int = Field(
        default=4,
        ge=1,
        le=8,
    )

    user_role: str = Field(
        default="standard_user",
        min_length=2,
        max_length=100,
    )


class CompiledParameterResponse(BaseModel):
    name: str
    value: Any
    data_type: str


class CompiledTableResponse(BaseModel):
    metadata_table_id: int
    data_source_id: int
    schema_name: str | None
    table_name: str
    alias: str


class CompiledJoinResponse(BaseModel):
    join_mapping_id: int
    relationship_id: int

    join_type: str

    source_table_alias: str
    source_column_name: str

    target_table_alias: str
    target_column_name: str

    expression: str


class SQLCompilerResponse(BaseModel):
    prompt: str

    decision: str
    is_compiled: bool

    dialect: SupportedDialect | None

    sql: str | None

    parameters: list[
        CompiledParameterResponse
    ]

    tables: list[
        CompiledTableResponse
    ]

    joins: list[
        CompiledJoinResponse
    ]

    approved_limit: int

    overall_confidence: int

    warnings: list[str]
    errors: list[str]
    explanation: list[str]

File: frontend/src/types/sqlCompiler.ts
export type SupportedSqlDialect =
  | "mysql"
  | "postgresql"
  | "oracle"
  | "mssql";

export interface SqlCompilerRequest {
  prompt: string;
  domain_id: number | null;
  requested_limit: number;
  maximum_entities: number;
  maximum_path_depth: number;
  user_role: string;
}

export interface CompiledParameter {
  name: string;
  value: unknown;
  data_type: string;
}

export interface CompiledTable {
  metadata_table_id: number;
  data_source_id: number;
  schema_name: string | null;
  table_name: string;
  alias: string;
}

export interface CompiledJoin {
  join_mapping_id: number;
  relationship_id: number;
  join_type: string;

  source_table_alias: string;
  source_column_name: string;

  target_table_alias: string;
  target_column_name: string;

  expression: string;
}

export interface SqlCompilerResult {
  prompt: string;

  decision: string;
  is_compiled: boolean;

  dialect: SupportedSqlDialect | null;

  sql: string | null;

  parameters: CompiledParameter[];
  tables: CompiledTable[];
  joins: CompiledJoin[];

  approved_limit: number;
  overall_confidence: number;

  warnings: string[];
  errors: string[];
  explanation: string[];
}
